src/Compilers/CSharp/Portable/Lowering/BoundTreeToDifferentEnclosingContextRewriter.cs

```diff
@@ -188,6 +188,12 @@ public override BoundNode VisitAwaitableValuePlaceholder(BoundAwaitableValuePlac
                     ;
         }
 
+        public override BoundNode? VisitMethodDefIndex(BoundMethodDefIndex node)
+        {
+            // Cannot replace a MethodDefIndex's Method/Type with a substituted symbol.
+            return node;
```

Comment:
This fixes the observed issue and causes the new test to pass in debug mode. It looks like the particular `Method` which was being substituted previously, was the original member method `AsAsyncEnumerable<T>`, which had its type parameters substituted with that from the display class `C.<AsAsyncEnumerable>d__0<T>`.

But this causes us to use a substituted method symbol in `EmitMethodDefIndexExpression`. I didn't debug further past that, but, it looks like that causes bogus index values to be used into the instrumentation payloads.